tox.ini

```diff
@@ -65,98 +70,27 @@ commands =
         -m pytest -v --tb native -s {posargs} {[vars]tests_path}/unit
     coverage report
 
-[testenv:charm-integration]
-description = Run charm integration tests
+[testenv:{j2,j3}-{charm,tls,h-scaling,ha,client}-integration]
+description = Run integration tests
 pass_env =
     {[testenv]pass_env}
     CI
     CI_PACKED_CHARMS
 deps =
+    # These are libjuju versions, not Juju.
+    j2: juju==2.9.42.1
+    j3: juju==3.1.0.1
```

Comment:
@Mehdi-Bendriss, which UX do you prefer? 

In terms of a single shared tox environment for each libjuju version, I think changing local juju versions isn't frequent right now but it might get in the way when we're planning juju 2 -> 3 upgrade testing. Not disagreeing with your point, but it may be annoying if we have to download it mid-test in the scenario where an integration test uses libjuju 2 to deploy the charm, but libjuju 3 to test the upgraded charm.

Reply:
Since you can call tox environments from one another, an option for automated upgrade testing would be to build the model using libjuju 2 in a `setup-integration` environment, upgrade juju, and then test the model in a second `upgrade-integration` tox environment that uses version 3. It clogs the tox file a bit, but it seems like a relatively clean way to not have to constantly reinstall the same package twice in a test.

Reply:
you add can tox to the [`allowlist-externals` field in the tox config](https://tox.wiki/en/latest/config.html#allowlist_externals), then just call `tox -e another-env` in the `commands` section.